internal/status/cache.go

```diff
@@ -38,22 +38,50 @@ const (
 func NewCache() Cache {
 	return Cache{
 		proxyUpdates: make(map[types.NamespacedName]*ProxyUpdate),
+		entries:      make(map[string]map[types.NamespacedName]CacheEntry),
 	}
 }
 
+type CacheEntry interface {
+	AsStatusUpdate() k8s.StatusUpdate
+	ConditionFor(ConditionType) *contour_api_v1.DetailedCondition
+}
+
 // Cache holds status updates from the DAG back towards Kubernetes.
 // It holds a per-Kind cache, and is intended to be accessed with a
 // KindAccessor.
 type Cache struct {
 	proxyUpdates map[types.NamespacedName]*ProxyUpdate
+
+	entries map[string]map[types.NamespacedName]CacheEntry
+}
+
+func (c *Cache) Acquire(obj k8s.Object) CacheEntry {
```

Comment:
`Acquire` and `Release` imply to me some kind of exclusive lock/lease semantics, which isn't the case here - maybe just `Get` and `Put` would be more accurate.

Reply:
These names emphasize the you are getting (possibly shared) access to an entry. Later on, we may need to deal with HTTPProxy using this in a reentrant pattern. "get" and "put" (at least to me), imply getting copies.

Reply:
+1 on Aquire & Release implying a lock/lease semantic. Contour doesn't use this terminology anywhere and I think it does make it less clear. The user shouldn't need to understand if the entry is shared or not and that logic should be hidden away in the implementation of the methods.

Reply:
It's for https://github.com/projectcontour/contour/issues/2968

Reply:
> +1 on Aquire & Release implying a lock/lease semantic. Contour doesn't use this terminology anywhere and I think it does make it less clear. The user shouldn't need to understand if the entry is shared or not and that logic should be hidden away in the implementation of the methods.

Something has to know that the cache entry can be shared. Either the cache needs to arbitrate access, the entry needs to arbitrate, or the caller needs to deal with it. In the specific case of `ExtensionService` we know that the caller can do that, and we also know that there is no reentrancy.

I'm cautiously optimistic that we can move HTTPProxy to use the generic cache, but that's a bigger change that needs more time and care.

I renamed the methods in the interest of moving this review forward.